### ISSUE TITLE
Proposal: Bounded Context Canvas V4

### ISSUE THREAD

**Issue by NTCoding** [ddd-crew/bounded-context-canvas#8]
One of the topics that frequently comes up is around the flow of messages and dependencies. The canvas can be improved in this aspect.

People like Vaughn Vernon have also mentioned that the canvas could show flow more. After leaving this on the backburner it now seems like the right way to go, and it will make the canvas clearer and almost self-documenting.

Messages flow in, business decisions are applied expressed in ubiquitous language, and messages go out.

So here's what I'm thinking:

![bcc](https://user-images.githubusercontent.com/692094/85196440-27dea380-b2d2-11ea-91d7-e5330a352478.jpg)

I don't this is 100% perfect, but I think it's an improvement.

Feedback and alternative proposals are always good. I've also started a twitter thread requesting feedback: https://twitter.com/ntcoding/status/1274243753440620544?s=20

**Comment by Max-Git** [comment]
Love it. Also this could sort of act like a C4-Like diagram where one could "click" on the central box and zoom in to an [aggregate design canvas](https://github.com/ddd-crew/aggregate-design-canvas) where stuff would be already pre-filled in (Invariants, Handled Commands, Created Events, ...) :)

**Comment by NTCoding** [comment]
@Max-Git: Software park are already working on such a tool which could support this purpose: https://github.com/Softwarepark/Contexture

**Comment by NTCoding** [comment]
Some feedback so far:

1. Model traits is not useful (remove it or make it clearer)
2. Messages is confusing because people think it means async messages not logical messages
3. Try to align messages with dependencies

So here's an attempt at addressing those issues:

![bcc2](https://user-images.githubusercontent.com/692094/85200629-0640e480-b2f1-11ea-92d5-e3912abd817a.jpg)

**Comment by yellowbrickc** [comment]
Hi @NTCoding, 
very cool, as always!

I know, I'm picky but I would flip Ubiquitous Language and Domain Roles. The Language "rules them all", the wording used for business rules and domain roles should follow the language.
What exactly are domain roles (or traits)?

**Comment by Max-Git** [comment]
@NTCoding from my point of view, I think a mix of both would be great. I'd keep the first one and change "Messages Consumed" &  "Messages Produced" with "Inbound Communication" & "Outbound Communication". Also changing "Model Traits" to "Domain Roles" is more clearer.

**Comment by kerry-perret** [comment]
n00b here

I'm bit confused, why query is no longer in the outbound communication?
![image](https://user-images.githubusercontent.com/11332444/85203807-790d8800-b310-11ea-9091-3460680757e7.png)
![image](https://user-images.githubusercontent.com/11332444/85203788-58453280-b310-11ea-9c66-6de7ab6935fc.png)

Actually, I can't really wrap my head around inbound and outbound communications, would it be possible to have an example between the two? Actually an example for each kind of block would probably help (ie. decision, command, query, event).

**Comment by NTCoding** [comment]
Hi @kerry-perret,

Thanks for pointing those issues out. Btw, n00b opinions are not only welcome but highly valued here.

Someone suggested on Twitter that it would be useful to show where the messages are coming from and going to. So I tried to spike up what that might look like. As you pointed out, my draft was very confusing, my intention was to read it in horizontal swim lanes like this:

![concept2](https://user-images.githubusercontent.com/692094/85207739-d6123980-b322-11ea-8b37-a41828b76dca.jpg)

If you have any suggestions or want to spike up an alternative design, I can only encourage you :)

**Comment by emgsilva** [comment]
@NTCoding another great iteration. Some remarks / food for thought:

- 1: I would rename ```Business Decisions``` section to ```Business Description``` or maybe ```Business Context```, or something else other than ```Business Decisions```, as it seems to capture only a subset of this part (namely the decisions, there will be other elements here).
- 2: I love the swimlanes per "collaborator". What I still miss is some semantic description of the actual "relationship". This could be, for example, the be patterns from context-mapping (maybe some remark linking to the Cheat Sheet from Michael). Reason for this is to make this swimlanes and collaborations descriptions even clearer, e.g.: our bounded context is conformist to the sales context... etc. I know we can do that explicitly on the Context Map diagram and avoid overload this image, but somehow having that "sugar" also here could be interesting to see the whole picture in a single canvas.
- 3: I have been thinking for a while how to also include here team topologies inter-relations. For "X-as-a-service" it will most likely be captured by relations I mention on point 2) - as this is about technical integration of BCs. However, for "Facilitation" and "Collaboration" it is not (in general) about technical communications and message exchange. I question whether it is relevant to have this on the BC Canvas (or maybe somewhere else), but if it was to include it here I would most likely include it underneath the core part (Business & Ubiquitous Language) - like "supporting/enabling" collaborations for the BC (and team) . We can think further about this, but I really think the BC canvas could provide the one pager that also shows this sort of "collaborations" (not only the technical integrations). (We may argue whether these are per BC or per Team, which is not always a 1-to-1, but I would like to pick your brain on that one :)).

Basically see 2) and 3) sketched in orange bellow:

![BC-Canvas-Extended-ContextMapsPatterns-TeamTopologies](https://user-images.githubusercontent.com/520307/85210202-3662a480-b33e-11ea-9be2-c94b65aa6eb0.png)

What do you think?

**Comment by skleanthous** [comment]
My thoughts:
1. `Business Decisions` works for me imho. It may list policies, and processes, etc. but the important part, are the decisions which are taken as an outcome. Naming it `Business Context` somewhat dilutes this.
2. I _fully_ agree with @yellowbrickc that swapping `Ubiquitous Language` with `Domain Roles` would be more beneficial.
3. In its current form, it seems like the canvas is very close to a snapshot of a process-level Event Storming, due to the emphasis on communication and decisions. I agree these need to be here, but I believe these need to be less pronounced. Personally, I'd find the bounded context canvas more useful if it recorded some of the deductions of the event storming sessions in the form of text or some tables. What I'd like to see more:
   - Boundary assumptions (why some responsibilities were allocated here vs some other context)
   - Some info on the inner working of the Bounded context; not to just document the policies and decisions (although this is my assumptions based on that they're in boxes), but describe them a bit
   - Some info about the runtime aspects of this BC, and if it is a service, some information about availability expectations
   - Consistency expectations for the internal processes

**Comment by NTCoding** [comment]
@emgsilva Nice idea with the context map patterns. I spiked it up and it looks good to me:

![cm](https://user-images.githubusercontent.com/692094/85218448-4c01b980-b392-11ea-860d-ed630df72f61.jpg)

**Comment by NTCoding** [comment]
@emgsilva @skleanthous I hear what you're saying about team topologies, decision logs, runtime characteristics, consistency, and some of the other details that are missing. The challenge here is not turning the canvas into a monolith that is off-putting to beginners.

@Max-Git has already floated the idea of something C4 style which works in layers. Another option could be to create an extension to the canvas or an advanced canvas.

So let's try and find a way to keep it beginner-friendly but also has a "power user" mode :)

**Comment by emgsilva** [comment]
Makes perfect sense @NTCoding... I actually like (modular) monoliths :p , but adding all these modules may indeed bee too off-putting for people who want to use this for simpler use cases.

Maybe looking forward we could indeed create an "extended" version for the people that want to take things further, something like: "Bounded Context Canvas - Extended" :D (I have recently used Team Canvas on a team activity and found they have a way to address this, namely they have a "basic" version - for whom wants to address the most basic concerns; and a "full" version with a few more elements. This could be way to go forward. The alternative is to outscope this to other "documents" as mentioned (but some of the points mentioned don't seem to yet fall into any other document)...

Cheers!

**Comment by Nagelfar** [comment]
I wouldn't bloat this 'Bounded Context Canvas' with too much additional content but rather come up with some additional documents as mentioned by @emgsilva.
For the canvas I see two target groups using it to get an overview about a bounded context:
1. People with little experience in DDD
2. DDD-'Experts' 

For 1) renaming `Messages` and `Model Traits` and creating swimlanes definitely improved readability.
They might not get a lot out of things like `Strategic classification` and the `Context Mapping patterns`, but I think they still get a great overview about a bounded context and adding more information will probably start to confuse them.

Even for 2) I'm not sure if e.g. adding the type of a collaborator (e.g. `frontend`, `bounded context`, `external service`) really adds benefit, or if this information could already be part of e.g. a 'technical focused document'
This document could explain the technical realisation of a bounded context, e.g
- used protocols for communication
- runtime characteristics
- deployment information
-  technical description of the connection towards a collaborator

**Comment by skleanthous** [comment]
I also like the idea of an advanced version of the canvas.

With my suggestion above, I wasn't attempting to move the canvas towards a more advanced audience though. I'm just comparing it with existing patterns and practices and trying to see how we could make this more valuable.

I see two values for the canvas:
1. Forcing you to think about some aspects, or at the very least, force you to be explicit about some decisions
1. As documentation (for people joining after some workshops, for people to be able to trace some decisions, etc)

My suggestions were mostly around the second value, as a lot of what I suggested should be already discussed during the collaborative workshops that needed to happen.

For documentation purposes, the existing version (v3) is good because it's nicely structured and gives some context, but otherwise, it's little more compared to a snapshot of the process-level event storming with some text to describe what's going on. I think we stand to benefit from recording some of the decisions we're making around the shape of the bounded context, decisions which are happening regardless of experience with DDD and are important in documentation.

**Comment by mploed** [comment]
I like the idea of picking up Vaughn's recommendation to add more flow. In my opinion this new version also fits nicley with the ideas behind having driving and driven adapters and a core with the domain mode. I really like this new version.

I also liked the recommendation of adding some context mapping patterns. However, I do see a danger of overcomplicating things for folks who are new to the DDD world or the canvas. In my opinion we should look at C4 as an inspiration here by offering a drill down possibility where we can add additional perspectives and details in different diagrams, canvases.

**Comment by emgsilva** [comment]
@mploed Indeed, C4-like approach (with multiple views/perspectives/concerns) may be the way to go to "bring things together"... at the moment all of the elements/canvases/diagrams are a bit disconnected (and consequently have some "redundant" elements). 

We may require some further thinking as of the reasonable levels... In my opinion the https://github.com/ddd-crew/ddd-starter-modelling-process is sort of already laying ground for this thinking, but some further alignment is probably needed.

While thinking about this I came to this 4 rough parts/groups (have 4 groups/canvas be great too => "DDD-C4" :D): 

- 1: Understanding and Aligning: Big Picture Event Storming, to discover domains and align business priorities
- 2: Strategic Architecture: understand the most interesting architecture, most likely Context Maps and some other related elements
- 3: Org Strategy and Org Design: make sure the right strategy and org design is setup, using things like Core Domain Charts, elements from Team Topologies and organizational checks to validate the "strategic architecture"
- 4: Tactical Architecture: where we zoom into each part (Bounded Context), and per BC we can use things like BC Canvas and maybe even Aggregate canvas (at a level deeper)... 

This is a rough sketch I had of this on a recent [presentation](https://speakerdeck.com/emgsilva/intro-to-sociotechnical-architecture-co-designing-technical-and-organizational-architecture-to-maximize-impact?slide=31): 

![image](https://user-images.githubusercontent.com/520307/85380426-37adf000-b53d-11ea-93d9-4f823d73c32b.png)

I think it is still challenging to "align and integrate" more the many different things we have on these phases and activities... but like we do in C4, you don't always need/use all the 4 diagrams, many times you use just one, etc... But Simon Brown made such a clear story of the focus of each part (and gives some flexibility), which makes C4 very usable, simple and successful. 

Anyways, really happy we are talking about this... maybe we can start elaborating on this on another thread, as this one was mostly to give feedback on the BC Canvas (maybe a candidate for a new repo/project of the ddd-crew?).

**Comment by cremich** [comment]
@emgsilva I really like the levels you sketched. This reflects 100% how I work in my projects and could be a good guidance during a project lifecycle. 

@NTCoding thanks for the v4. I was wondering if the swimlanes are only present in the inbound/outbound sections. Would it also improve the readability if the swimlanes are also visible inside the "Business decision" box? This could help to map the decisions to a collaborator.

**Comment by NTCoding** [comment]
Sorry for the delay. Here is a shared miro board https://miro.com/app/board/o9J_kpBQ-jE=/

**Comment by NTCoding** [comment]
Hey everyone. I've created a poll with options for some of the things we discussed. I'll leave the poll open for a week and then we'll create a final draft V4 of the canvas with the winning choices. We'll then have a final round of discussion.

https://forms.gle/aMGEgk4JpRu1FCDLA

**Comment by NTCoding** [comment]
Here are the results of the survey. Thank you to everyone who contributed.

The most popular solutions are clear. Model traits has some interesting discussion though.

I'll go ahead and create a draft V4, but please feel free to discuss any of the proposed changes further or suggest new ideas. It's not too late for any big or small ideas to be discussed.

[Bounded Context Canvas V4 Proposed Changes.pdf](https://github.com/ddd-crew/bounded-context-canvas/files/4988387/Bounded.Context.Canvas.V4.Proposed.Changes.pdf)

**Comment by NTCoding** [comment]
![canvas_v4_empty](https://user-images.githubusercontent.com/692094/88673429-815f9c80-d0e0-11ea-9648-77abc01d1a85.jpg)

![virtual_template](https://user-images.githubusercontent.com/692094/88673448-87ee1400-d0e0-11ea-9332-827dac1e1dba.jpg)

**Comment by Baasie** [comment]
Looks great @NTCoding , one thing I might change, being maybe rather late is to change the Business Decision wording to Business Contraints. I like Domain Roles as a name better here.

I will make another issue for my acceptance criteria option I have been bugging you about for a while ;)

**Comment by NTCoding** [comment]
@Baasie I don't think we have to keep the name "Business Decisions". If there is something better we can change.

My initial feeling is that "constraints" might be too expert level for beginners. I think "Business Rules" is the obvious one that people know about but might be too specific so I thought "Business Decisions" might be clear but more generic.

How do you think we could make "Business Constraints" easy to understand for newcomers? Or maybe they already would and my perception is wrong.

**Comment by emgsilva** [comment]
@NTCoding like @Baasie I also (mentioned it before and) still think "Business Decisions" as title may not be "general" enough to represent the whole section, given that we also want to capture other elements, namely: "constraints" and "policies" specified/defined in the domain. We can argue "these are all decisions" we take in the for the bounded-context, but not sure if that is always the case (e.g.: a policy may just be something we need to comply, no way to go around it...).

What do you think about simply making the title: "Decisions, Policies & Rules" => this should make it very explicit and clear. We could still keep the "sub-title" with the same text "Key business decisions, policies and rules" (to stress the "Business" aspect). Not 100% convinced with such a long text... but could be a way to mitigate some of the points we are seeing.

**Comment by Baasie** [comment]
Good point @emgsilva , I still think constraints are a good name that is not too general. Because you have multiple kinds of constraints right? I also want to avoid lingo, it creates and in/outsider cultures and then again how well know is constraint?

So perhaps we can go with Constraints: "Business Decisions, Policies & Rules" Might be my proposal

**Comment by NTCoding** [comment]
I'm enjoying this discussion and I think we can simplify and improve the decisions part of the canvas so please keep sharing ideas. There is no rush to make a decisions.

My feeling is that "Constraints: "Business Decisions, Policies & Rules" is a bit too long. How could we shorten it. 

How about: Constraints, Rules, and Policies ? I feel like the "business" part is implied and it might be ok to drop it for something which is more explanatory.

**Comment by emgsilva** [comment]
@NTCoding I mentioned above, my impression is that "Decisions, Policies & Rules" could be a good title (still a bit big, but I think will fit the space :p). Furthermore it is "snappy to read" and I personally am more in favor of Decisions than Constraints, they are more widely understood - in my opinion - and I think we literally want to show "decisions" + "policies" + "rules" right?

**Comment by yellowbrickc** [comment]
I think the word "Constraints" is not too hard to understand (it is a normal word indicating some boundaries) so I would go for this combi:

** Business Constraints
*** Decisions, Policies, Rules

I am convinced, that nothing happens "just like that". The best example is GDPR: it is the law but still, a lot of companies decide to postpone its application. So all entries in this place would be some kind of hard guarding fences, obligations = constraints.

**Comment by NTCoding** [comment]
@yellowbrickc constraints is a popular word which people will understand, but in this particular I don't think it will be entirely clear what we mean. I can see a few interpretations "constraints of the implementation", "constraints enforced", etc. 

Without a qualifier, I also don't believe that most people will naturally associate constraints with business rules and policies.

I consider the word constraints (without any qualifier) to be a more precise word suitable for experts who understand it's definition in this context, but less clear for people outside of DDD. And I'm possibly wrong on this. 

I think constraints would need to be qualified somehow, e.g.:

- Business Constraints Enforced
- Business Constraints and Rules

**Comment by Baasie** [comment]
I think constraint would work well. We cannot fully avoid lingo, and also the lingo might help people think more about what is exactly meant with this word. I liked @yellowbrickc her business constraints, decisions, policies, rules.

Also going in on domain traits, I think it is what Rebecca Wirfs-brocks calls stereotypes in responsiblity driven design no?

**Comment by NTCoding** [comment]
@Baasie constraints is a very-overloaded term and I'm not sure what it means here. For example, how does a constraint relate to a business rule and a policy? Is it a superset or a sibling. I'm not opposed to the idea I think first we would need a clear definition with examples (maybe you have already defined them somewhere).

When I think about constraints in the non-generic sense, I think to the world of complexity where Dave Snowden talks about enabling and governing constraints. In both cases the word constraints has a qualifier to add a context-specific definition which is lacking here. In this case it could mean Implementation constraints or just general constraints that don't live as code inside the bounded context.

Business constraints is a qualifier but it already has popular usage focused more on business models. An example of a business constraint is "we don't have enough time to deliver this new product" or "we don't have the expertise in the company to make this strategy work".

I am still open to the idea of using constraints if we can define it and qualify it. Perhaps "Domain Constraints" might be a starting point?

**Comment by NTCoding** [comment]
@Baasie The Domain Roles links to concepts suggested by a few different people. Alberto calls some of them archetypes, traits are from psychology describing an aspect of someone's behaviour, and roles aligns with Rebecca's responsibility-driven design. She does use the word stereotype which I believe describes a cluster of roles.

In DDD, I don't think we have done much work to discover bounded context stereotypes yet, so if we focus on the roles we will start to see clusters and then we can start to identify bounded context stereotypes.

^^^ This is just my take on it by the way. Always good to get other perspectives.

**Comment by yellowbrickc** [comment]
@NTCoding "Always good to get other perspectives." => I think, we all agree on this. 
This discussion is going on for 55 days now. Let's make a decision and publish it to get those other perspectives from other people not in this circle. It must not be for all times, and we can improve it later IF it is necessary. Anything else is just guessing IMO

**Comment by NTCoding** [comment]
Personally I'm not too worried about the timelines. I feel that publishing new versions of the canvas cannot be done too frequently so more up-front design is necessary.


At present there is 1 discussion regarding business decisions. So there are 2 choices I can see:

1. We take the time to define "constraints" with a short description and examples and explaining how it relates to business rules and policies (this will be needed for the official readme in any case).

2. We publish the current draft and take more time to explore how the business decisions section can be improved and publish an official new version in a few months.

I am happy with both approaches. @yellowbrickc is your feeling that option 1 is the way to go for now?

**Comment by yellowbrickc** [comment]
actually I'm rather for option 2. I understand the need for the bigger upfront discussion but I think we exhausted that. This discussion started 12 days ago and nobody seems to have left anything to add. This is the reason why I think, it should be published. I have no problem with calling it a draft (everything is a current snapshot based on our "best shot" at a given time)
It is not the timeline what bothers me but not making any progress visible and useable by others.

**Comment by NTCoding** [comment]
@yellowbrickc I'm with you! I'll prepare the pull request of V4. 

In the future, maybe as a meetup event, we can explore the decisions/rules/policies/constraints section.

**Comment by cremich** [comment]
@NTCoding as I am actively using this version right now, I recognize that I do not completly understand the meaning of the relationship type that is defined on the ubiquitous language box. The one on the collaborator side is clear for me. It defines the relationship between our BC and the collaborator. But what does a relationship type on the UL box mean?

**Comment by NTCoding** [comment]
Hey @cremich. 

Some patterns, like partnership (PNR in the diagram below) describe both ends of a relationship. Whereas some patterns like Open Host (OHS) and Conformist (CF) describe only one side of the relationship. An OHS describes how something is exposed, while a CF describes how something is received.

Here's an example: 

![image](https://user-images.githubusercontent.com/692094/98363455-ec493980-2026-11eb-9c5d-631d8f565417.png)

In this diagram, the bounded context we are visualising (in the middle of the diagram) is exposing it's services via an OHS (Open Host Service). The "Find Nearest Showroom" is a standard API that multiple consumers might all use in the same way. The Company Website decides to conform (CF) to the Open Host, but other collaborators could instead translate (ACL) the Open Host.

With the row below that, a partnership is used. A Partnership (PNR) says that both teams are working together collaboratively, so describes both ends of the relationship. But here's where things get interesting...

A partnership states that two teams are working together towards a shared goal. It doesn't describe how the systems integrate technically. So the partnership could also have an OHS -> CF relationship.

@mploed articulates this perfectly with his table below, showing that different context mapping patterns address different aspects of the socio-technical relationships.


<img width="576" alt="Capture d’écran 2020-11-06 à 12 04 21" src="https://user-images.githubusercontent.com/692094/98364175-3aab0800-2028-11eb-81da-a39d4228f3e8.png">

**Comment by cremich** [comment]
Perfect explanation! Thank you @NTCoding
